KDOP_tree/include/CGAL/KDOP_tree/internal/KDOP_node.h

```diff
@@ -0,0 +1,388 @@
+    Primitive& right_data() { return *static_cast<Primitive*>(m_p_right_child); }
+
+  private:
+    // node bounding box
+    Bounding_box m_bbox;
+
+    // node kdop
+    Kdop m_kdop;
+    Vec_direction m_directions;
+
+    int m_direction_number;
+
+    // children nodes, either pointing towards children (if children are not leaves),
+    // or pointing toward input primitives (if children are leaves)
+    void *m_p_left_child;
+    void *m_p_right_child;
+
+  }; // end class KDOP_node
+
+  template<typename Tr>
+  template<typename ConstPrimitiveIterator>
+  void
+  KDOP_node<Tr>::expand(ConstPrimitiveIterator first,
+                        ConstPrimitiveIterator beyond,
+                        const std::size_t range,
+                        const Tr& traits)
+  {
+    // binary splitting as AABB
+    m_bbox = traits.compute_bbox_object()(first, beyond);
+
+    traits.split_primitives_object()(first, beyond, m_bbox);
+
+    switch(range)
+    {
+    case 2:
+      m_p_left_child = &(*first);
+      m_p_right_child = &(*(++first));
+      break;
+    case 3:
+      m_p_left_child = &(*first);
+      m_p_right_child = static_cast<Node*>(this) + 1;
+      right_child().expand(first + 1, beyond, 2, traits);
+      break;
+    default:
+      const std::size_t new_range = range/2;
+      m_p_left_child = static_cast<Node*>(this) + 1;
+      m_p_right_child = static_cast<Node*>(this) + new_range;
+      left_child().expand(first, first + new_range, new_range, traits);
+      right_child().expand(first + new_range, beyond, range - new_range, traits);
+    }
+  }
+
+#ifdef TEST_
+  template<typename Tr>
+  template<typename ConstPrimitiveIterator>
+  void
+  KDOP_node<Tr>::expand(ConstPrimitiveIterator first,
+                        ConstPrimitiveIterator beyond,
+                        const std::size_t range,
+                        const Tr& traits)
+  {
+    // binary splitting as AABB
+    traits.split_primitives_object()(first, beyond);
+
+    switch(range)
+    {
+    case 2:
+      m_p_left_child = &(*first);
+      m_p_right_child = &(*(++first));
+      break;
+    case 3:
+      m_p_left_child = &(*first);
+      m_p_right_child = static_cast<Node*>(this) + 1;
+      right_child().expand(first + 1, beyond, 2, traits);
+      break;
+    default:
+      const std::size_t new_range = range/2;
+      m_p_left_child = static_cast<Node*>(this) + 1;
+      m_p_right_child = static_cast<Node*>(this) + new_range;
+      left_child().expand(first, first + new_range, new_range, traits);
+      right_child().expand(first + new_range, beyond, range - new_range, traits);
+    }
+  }
+#endif
+
+  template<typename Tr>
+  void
+  KDOP_node<Tr>::union_support_heights(const Array_height& left_heights,
+                                       const Array_height& right_heights,
+                                       Array_height& heights_union,
+                                       const int direction_number)
```

Comment:
No need for `direction_number`.  Should be an `unsigned int` template parameter.  Maybe go through the entire code to clean that up in a single commit.